docs/proposals/kubectl-apply-purge-from-namespace.md

```diff
@@ -0,0 +1,103 @@
+
+All resources referred to must be explicitly labelled with one non-default namespace in order for the `--purge-from-namespace` argument to succeed.
+
+This flag will require `--namespace`, and will not accept namespaces set in the input resource definitions.
+
+There shall exist a well-known annotation (`kubectl.kubernetes.io/purge-from-namespace=false`) that can be used to exclude some resource(s) from being purged.
+
+The goal here is to avoid a negative user experience where the user accidentally deletes all the resources in some other namespace just because they accidentally refer to (e.g.) a single resource in some other namespace.
+
+## Managing Add-ons
+
+Let's first consider the add-ons use case.
+
+There exists a `/etc/kubernetes/addons` directory with some number of files. All resources declared in those files belong to the same namespace (`kube-system`), which is set explicitly in each of the resources.
+
+When `kubectl apply --purge-from-namespace -f /etc/kubernetes/addons` runs for the first time all the resources will be created.
+
+Eventually some new files appear in `/etc/kubernetes/addons` and some are modified.
+
+Subsequent invocations of `kubectl apply --purge-from-namespace -f /etc/kubernetes/addons` result in new resources being created and modified resources being updated.
+
+When user deletes some of the files or resources within files from `/etc/kubernetes/addons`, `kubectl apply --purge-from-namespace` would ensure the resources that were referenced in the deleted files are deleted, based on the assumption that `/etc/kubernetes/addons` represents all resources that are supposed to be present in `kube-system` namespace.
+
+
+## Managing Any Other Application
+
+It should be possible to apply the same approach to any other applications, with the assumption that single application occupies a namespace of its own.
+
+The usage may look like this:
+
+```
+kubectl apply --purge-from-namespace -f https://example.com/myapp.json
+```
+
+Given resources in `myapp.json` set their namespaces explicitly, resource that are no longer in this file would be purged, if no longer defined.
+
+```
+kubectl apply --purge-from-namespace -f https://example.com/myapp-part1.json -f https://example.com/myapp-part2.json
+```
+
+If `myapp-part1.json` and `myapp-part2.json` are in the same namespaces and it is explicitly defined, removing either of the URLs from the invocations arguments will result in all the resources that URL has defined to be purged.
+
+## Flag Discovery User Experience
+
+As stated above, the `--purge-from-namespace` flag shouldn't be enabled by default, however it should be discoverable.
+
+So, if a `kubectl apply` command is run without `--purge-from-namespace`, and the referenced files are all in the same namespace, then the command could output something like:
```

Comment:
I'm not sure how keen I am to do this. It will make apply significantly more expensive (to discover all available resources within the namespace) and make the user experience significantly different than other forms of updates. 

`--dry-run`, to simulate what changes would be done, would be useful.

Mentioning this in the short help snippet for kubectl apply would be useful.

A tutorial that shows how to manage resources using apply would be useful.